### PR TITLE
Fix charged and neutral candidate converters for reco::Jets

### PR DESCRIPTION
**Pull request by SWuchterl**
#### PR description:

This PR allows the evaluation of DeepJet on `reco::Jet`s, which to my understanding should have worked before the latest change. With the current logic, only `pat::Jet`s are accepted.
The fix is needed to use DeepJet at HLT. For the `reco::` case, features needed for DeepDoubleX are filled with the same dummy variables as in the case of no subjets.

#### PR validation:
Code compiles.
`40 39 38 29 18 4 1 1 1 tests passed, 0 0 0 0 0 0 0 0 0 failed`
Also, tested on data and MC samples, running a modified HLT menu with DeepJet instead of DeepCSV.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
N/A

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
The IB is broken.

tests will need to be restarted after the new IB shows up anyways to get appropriate comparisons

File: RecoBTag/FeatureTools/interface/ChargedCandidateConverter.h [hunk 1]
```diff
@@ -53,20 +53,22 @@ namespace btagbtvdeep {
...
-      c_pf_features.drsubjet1 = !subjets.empty() ? reco::deltaR(*c_pf, *subjets.at(0)) : -1;
-      c_pf_features.drsubjet2 = subjets.size() > 1 ? reco::deltaR(*c_pf, *subjets.at(1)) : -1;
+    if (patJet) {
+      if (patJet->nSubjetCollections() > 0) {
+        auto subjets = patJet->subjets();
+        // sort by pt
+        std::sort(subjets.begin(), subjets.end(), [](const edm::Ptr<pat::Jet>& p1, const edm::Ptr<pat::Jet>& p2) {
+          return p1->pt() > p2->pt();
+        });
+        c_pf_features.drsubjet1 = !subjets.empty() ? reco::deltaR(*c_pf, *subjets.at(0)) : -1;
```
Comment (VinInn) [inline RecoBTag/FeatureTools/interface/ChargedCandidateConverter.h:64]: use ```operator[]``` not ```at```

File: RecoBTag/FeatureTools/interface/NeutralCandidateConverter.h [hunk 2]
```diff
@@ -32,18 +32,20 @@ namespace btagbtvdeep {
...
     // Do Subjets
-    if (patJet->nSubjetCollections() > 0) {
-      auto subjets = patJet->subjets();
-      // sort by pt
-      std::sort(subjets.begin(), subjets.end(), [](const edm::Ptr<pat::Jet>& p1, const edm::Ptr<pat::Jet>& p2) {
-        return p1->pt() > p2->pt();
-      });
-      n_pf_features.drsubjet1 = !subjets.empty() ? reco::deltaR(*n_pf, *subjets.at(0)) : -1;
-      n_pf_features.drsubjet2 = subjets.size() > 1 ? reco::deltaR(*n_pf, *subjets.at(1)) : -1;
+    if (patJet) {
```
Comment (VinInn) [inline RecoBTag/FeatureTools/interface/NeutralCandidateConverter.h:36]: why all this code duplication?

File: RecoBTag/FeatureTools/interface/ChargedCandidateConverter.h [hunk 3]
```diff
@@ -53,20 +53,22 @@ namespace btagbtvdeep {
...
-      std::sort(subjets.begin(), subjets.end(), [](const edm::Ptr<pat::Jet>& p1, const edm::Ptr<pat::Jet>& p2) {
-        return p1->pt() > p2->pt();
-      });
-      c_pf_features.drsubjet1 = !subjets.empty() ? reco::deltaR(*c_pf, *subjets.at(0)) : -1;
-      c_pf_features.drsubjet2 = subjets.size() > 1 ? reco::deltaR(*c_pf, *subjets.at(1)) : -1;
+    if (patJet) {
+      if (patJet->nSubjetCollections() > 0) {
+        auto subjets = patJet->subjets();
+        // sort by pt
+        std::sort(subjets.begin(), subjets.end(), [](const edm::Ptr<pat::Jet>& p1, const edm::Ptr<pat::Jet>& p2) {
```
[the remark this answers is not in the document]
Comment (SWuchterl) [reply RecoBTag/FeatureTools/interface/ChargedCandidateConverter.h:61]: Those pieces of the code were already there, I just inverted the logic of the if statement in the beginning. From the diff it seems different, I know. But of course, I can change it as suggested.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoBTag/FeatureTools/interface/NeutralCandidateConverter.h (hunk 2 above)
Comment (SWuchterl) [reply RecoBTag/FeatureTools/interface/NeutralCandidateConverter.h:36]: Could you please clarify the comment about what code exactly is duplicated in this file?

File: RecoBTag/FeatureTools/interface/ChargedCandidateConverter.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoBTag/FeatureTools/interface/ChargedCandidateConverter.h:61]: stil this code is very sub-optimal

File: RecoBTag/FeatureTools/interface/NeutralCandidateConverter.h (hunk 2 above)
Comment (VinInn) [reply RecoBTag/FeatureTools/interface/NeutralCandidateConverter.h:36]: the code in this file looks identical to the one in the previous one.

Comment (AlexDeMoor) [reply RecoBTag/FeatureTools/interface/NeutralCandidateConverter.h:36]: The first file is about Charged candidates and this one is about the Neutral ones. Some elements look similar indeed but they define 2 different inputs used by the neural network.

**Review by jpata — COMMENTED** [review]
There is some duplication and a possible code optimization that was brought up by Vincenzo that could be addressed. I don't see any reco changes - good.
After changes are implemented, we could run another round of tests with profiling.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Thanks for the updates. It looks like there are unexpected reco differences now. Can you check with some printouts if the max_elements logic is giving the expected result?

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-03-2300+4a1543/46661/validateJR/all_mini_OldVSNew_Higgs200ChargedTauswf9p0/all_mini_OldVSNew_Higgs200ChargedTauswf9p0c_min2,max-2,patJets_slimmedJetsAK8__RECO_obj___pairDiscriVector__52__second.png

**Comment by SWuchterl** [comment]
> Thanks for the updates. It looks like there are unexpected reco differences now. Can you check with some printouts if the max_elements logic is giving the expected result?

You are right. Although the search should have worked, something went wrong with the comparison in the lambda function. I now replaced it with `nth_element` which only sorts the first two maximum entries of the vector.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
+reconstruction
- fix for using DeepJet at HLT by setting the subjet features.
- no reco changes expected or observed

**Comment by qliphy** [comment]
+1

**Pull request merged**